lib/cloud/azure/imds.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, client.baseURL+route, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	req.Header.Add("Metadata", "True")
+	query := req.URL.Query()
+	query.Add("format", "json")
+	if client.apiVersion != "" {
+		query.Add("api-version", client.apiVersion)
+	}
+	req.URL.RawQuery = query.Encode()
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	body, err := utils.ReadAtMost(resp.Body, teleport.MaxHTTPResponseSize)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, parseMetadataClientError(resp.StatusCode, body)
+	}
+	return body, nil
+}
+
+// getVersions gets a list of all API versions supported by this instance.
+func (client *InstanceMetadataClient) getVersions(ctx context.Context) ([]string, error) {
+	versions := struct {
+		APIVersions []string `json:"apiVersions"`
+	}{}
+	body, err := client.getRawMetadata(ctx, "/versions")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := utils.FastUnmarshal(body, &versions); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return versions.APIVersions, nil
+}
+
+// IsAvailable checks if instance metadata is available.
+func (client *InstanceMetadataClient) IsAvailable(ctx context.Context) bool {
+	if client.apiVersion != "" {
+		return true
+	}
+
+	err := client.selectVersion(ctx)
+	return err == nil
+}
+
+// GetTags gets all of the Azure instance's tags.
+func (client *InstanceMetadataClient) GetTags(ctx context.Context) (map[string]string, error) {
+	if !client.IsAvailable(ctx) {
+		return nil, trace.NotFound("Instance metadata not available")
+	}
+
+	rawTags := []struct {
+		Name  string `json:"name"`
+		Value string `json:"value"`
+	}{}
+	body, err := client.getRawMetadata(ctx, "/instance/compute/tagsList")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := utils.FastUnmarshal(body, &rawTags); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tags := make(map[string]string, len(rawTags))
+	for _, tag := range rawTags {
+		tags[tag.Name] = tag.Value
+	}
+	return tags, nil
+}
+
+// GetHostname gets the hostname set by the cloud instance that Teleport
+// should use, if any.
+func (client *InstanceMetadataClient) GetHostname(ctx context.Context) (string, error) {
+	tags, err := client.GetTags(ctx)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	value, ok := tags[types.CloudHostnameTag]
+	if !ok {
+		return "", trace.NotFound("Tag %q not found", types.CloudHostnameTag)
+	}
+	return value, nil
+}
+
+// selectVersion selects the most recent API version greater than or equal to
+// a minimum version. Versions are represented as dates of the form YYYY-MM-DD.
+func selectVersion(versions []string, minimumVersion string) (string, error) {
+	if len(versions) == 0 {
+		return "", trace.BadParameter("No versions provided")
```

Comment:
Can this error message be made more helpful? E.g. how/when can this happen?

Also, returned error messages should start with lowercase.